newapi/ooniapi/auth.py

```diff
@@ -356,16 +366,23 @@ def user_login():
 def _set_account_role(email_address, role: str) -> int:
     account_id = hash_email_address(email_address)
     # log.info(f"Giving account {account_id} role {role}")
-    query = """INSERT INTO accounts (account_id, role)
-        VALUES(:account_id, :role)
-        ON CONFLICT (account_id) DO
-        UPDATE SET role = EXCLUDED.role
-    """
     # TODO: when role is changed enforce token expunge
     query_params = dict(account_id=account_id, role=role)
-    q = current_app.db_session.execute(query, query_params).rowcount
-    current_app.db_session.commit()
-    # TODO: return update/insert count
+    if current_app.config["DATABASE_URI_RO"]:
+        query = """INSERT INTO accounts (account_id, role)
+            VALUES(:account_id, :role)
+            ON CONFLICT (account_id) DO
+            UPDATE SET role = EXCLUDED.role
+        """
+        q = current_app.db_session.execute(query, query_params).rowcount
+        current_app.db_session.commit()
+
+    if current_app.config["USE_CLICKHOUSE"]:
+        query = """INSERT INTO accounts (account_id, role)
+            VALUES(:account_id, :role)"""
```

Comment:
I guess the `accounts` table in clickhouse is configured to be a `ReplacingMergeTree`?

If so, isn't there the risk that after the "update" of a role is executed we end up with duplicate rows and hence the changes to the account role will not be reflected immediately?

Perhaps we can run an `OPTIMIZE` query right after this (though this feels a bit like a hack)?

Reply:
Alternatively we could make use the table mutations: https://clickhouse.com/docs/en/sql-reference/statements/alter/update/

Reply:
No, these are EmbeddedRocksDB tables. A little database embedded in Clickhouse that provides ACID.